### DISCUSSION TITLE
Proposal: allow fixed buffer directly in class

### DISCUSSION THREAD

**Discussion by ygc369** [dotnet/csharplang#3295] [Language Ideas]
According to fixed buffer proposal (https://github.com/dotnet/csharplang/blob/master/proposals/fixed-sized-buffers.md), the fixed buffer can only be in a struct, if we want to use it in a class, we must use a struct to wrap it.
For example,
```C#
struct A
{
    public fixed object a[10];
}
class B
{
    A buffer;
}
//if we want to access the fixed buffer, we have to use the following code:
var b = new B();
b.buffer.a[0]=22;
```
But if fixed buffer can directly be in a class, then the code above can be simplified as following:
```C#
class B
{
    public fixed object a[10];
}
var b = new B();
b.a[0]=22;
```

**Comment by Unknown6656** [comment]
Then why use a fixed buffer at all? Why not use
```csharp
class B
{
    public static readonly object[] a = new object[10];
}

var b = new B();
b.a[0]=22;
```

Fixed buffers are usually only used in combination with P/Invoke and other native interaction.

**Reply by sgf** [reply]
because we need the friendly struct layout in the memory(direct read and write)

**Comment by ygc369** [comment]
@Unknown6656 
Your example has to do two heap allocations, one is `class B` itself, the other is the array inside it. But if fixed buffer could be used here, only one heap allocation is needed. Thus GC pressure can be reduced. Besides, when accessing the array element, the fixed buffer might provide better performance, for it only need to do dereference once, but in your example, dereference need to be done twice.

**Comment by AartBluestoke** [comment]
the bigger effect is probably the memory co-locality and potential cpu cache-miss on de-reference, as if b and its contents are alloc'd separately they will probably end up on different cache lines. Runtime cpu performance is often limited by cache access (even if it looks cpu-compute bound).

**Comment by ygc369** [comment]
We can use a struct to wrap fixed buffer, but that would introduce a new type into our assembly that only a wrapper. It then exposed in reflection. And if we use this pattern in many classes then our assembly will be bloated with useless struct of the same pattern. So if we can place fixed buffer directly in class, then we can avoid metadata bloating.

**Comment by john-h-k** [comment]
Yeah, i think this would be useful in a set of niche situations and has basically no drawbacks
